clash-lib/src/Clash/Normalize/PrimitiveReductions.hs

```diff
@@ -929,14 +932,41 @@ reduceUnconcat n 0 aTy arg = do
       | (Just vecTc)     <- lookupUniqMap vecTcNm tcm
       , nameOcc vecTcNm == "Clash.Sized.Vector.Vec"
       , [nilCon,consCon] <- tyConDataCons vecTc
-      = let nilVec           = mkVec nilCon consCon aTy 0 []
-            innerVecTy       = mkTyConApp vecTcNm [LitTy (NumTy 0), aTy]
-            retVec           = mkVec nilCon consCon innerVecTy n (replicate (fromInteger n) nilVec)
-        in  changed retVec
+      , let innerVecTy = mkTyConApp vecTcNm [LitTy (NumTy m), aTy]
+      = if n == 0 then
+          changed (mkVecNil nilCon innerVecTy)
+        else if m == 0 then
+          let nilVec           = mkVecNil nilCon aTy
+              retVec           = mkVec nilCon consCon innerVecTy n
+                                   (replicate (fromInteger n) nilVec)
+          in  changed retVec
+        else do
+          uniqs0 <- Lens.use uniqSupply
+          let (uniqs1,(vars,elems)) = second (second concat . unzip)
+                                    $ extractElems uniqs0 inScope consCon aTy
+                                       'U' (n*m) arg
+              -- Build a vector out of the first m elements
+              mvec = mkVec nilCon consCon aTy m (take (fromInteger m) vars)
+              -- Get the vector representing the next ((n-1)*m) elements
+              (lbs,head -> nextVec) = splitAt ((2*fromInteger m)-1) elems
```

Comment:
I don't understand what's happening here, shouldn't this split at `m`?

Reply:
No, `extractElems` creates 2*N bindings for a vector of `N` long: so for `xs :: Vec 4 a` it creates:
```haskell
x0 = head xs
xs0 = tail xs
x1 = head xs0
xs1 = tail xs0
x2 = head xs1
xs2 = tail xs1
x3 = head xs2
xs3 = tail xs2
```

Reply:
Improved the comments